python/pyspark/sql/udf.py

```diff
@@ -157,7 +157,17 @@ def _create_judf(self):
         spark = SparkSession.builder.getOrCreate()
         sc = spark.sparkContext
 
-        wrapped_func = _wrap_function(sc, self.func, self.returnType)
+        func = fail_on_stopiteration(self.func)
+
+        # for pandas UDFs the worker needs to know if the function takes
+        # one or two arguments, but the signature is lost when wrapping with
+        # fail_on_stopiteration, so we store it here
+        if self.evalType in (PythonEvalType.SQL_SCALAR_PANDAS_UDF,
+                             PythonEvalType.SQL_GROUPED_MAP_PANDAS_UDF,
+                             PythonEvalType.SQL_GROUPED_AGG_PANDAS_UDF):
+            func._argspec = _get_argspec(self.func)
```

Comment:
Does it make sense for `fail_one_stopiteration`  to keep the function signature instead of restoring them here?

Reply:
I am not sure how to do that, though, can you suggest a way? Originally this hack was in `fail_one_stopiteration`, but then it was decided to restrict its scope as much as possible

Reply:
That seems to be difficult in particular in Python 2. I'm not aware of a clean and straight forward way without a hack to copy the signature in Python 2.

Reply:
Yup, we definitely support. The current approach probably wouldn't change anything we supported before. I believe the builtin functions in Python 2 don't already with with Pandas UDFs.

Reply:
I see. Thanks for the clarification.